pkg/metadata/manager.go

```diff
@@ -0,0 +1,374 @@
+	// Channel size for pod events in order to not block callers
+	podEventsChannelSize = 20
+)
+
+// CgroupManager maintains Kubernetes and low-level metadata (cgroup path and
+// cgroup id) for local pods and their containers. In order to do that, it defines
+// and implements callback functions that are called on Kubernetes pod watcher events.
+// It also exposes APIs to read the saved metadata.
+//
+// The manager's internals are synchronized via a channel, and must not be
+// accessed/updated outside this channel.
+//
+// During initialization, the manager checks for a valid base cgroup path.
+// In case of environments using non-default cgroup base paths, manager will
+// fail to get a valid cgroup base path, and ignore all the subsequent pod events.
+type CgroupManager struct {
+	// Map of pod metadata indexed by their UIDs
+	podMetadataById map[podUID]*podMetadata
+	// Map of container metadata indexed by their cgroup ids
+	containerMetadataByCgrpId map[uint64]*containerMetadata
+	// Set to the valid cgroup base path if found
+	templateCgroupBasePath string
+	// Buffered channel to receive pod events
+	podEvents chan podEvent
+	// Object to check cgroup base path
+	checkCgroupPath *sync.Once
+	// Flag to check if manager is enabled, and processing events
+	enabled bool
+	// Channel to shut down manager
+	shutdown chan struct{}
+	// Interface to do file operations
+	fschecker fs
+	// Interface to do cgroups related operations
+	cgroupsChecker cgroup
+}
+
+type PodMetadata struct {
+	Name      string
+	Namespace string
+	IPs       []string
+}
+
+const (
+	PodAddEvent = iota
+	PodUpdateEvent
+	PodDeleteEvent
+	PodMetadataEvent
+)
+
+func NewCgroupManager() *CgroupManager {
+	m := &CgroupManager{
+		podMetadataById:           make(map[string]*podMetadata),
+		containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
+		podEvents:                 make(chan podEvent, podEventsChannelSize),
+		shutdown:                  make(chan struct{}),
+	}
+	m.fschecker = fsImpl{}
+	m.cgroupsChecker = cgroupImpl{}
+	m.checkCgroupPath = new(sync.Once)
+
+	m.enable()
+
+	return m
+}
+
+func (m *CgroupManager) OnAddPod(pod *v1.Pod) {
+	if pod.Spec.NodeName != nodetypes.GetName() {
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       pod,
+		eventType: PodAddEvent,
+	}
+}
+
+func (m *CgroupManager) OnUpdatePod(oldPod, newPod *v1.Pod) {
+	if newPod.Spec.NodeName != nodetypes.GetName() {
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       newPod,
+		oldPod:    oldPod,
+		eventType: PodUpdateEvent,
+	}
+}
+
+func (m *CgroupManager) OnDeletePod(pod *v1.Pod) {
+	if pod.Spec.NodeName != nodetypes.GetName() {
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       pod,
+		eventType: PodDeleteEvent,
+	}
+}
+
+// GetParentPodMetadata returns parent pod metadata for the given container
+// cgroup id in case of success, or nil otherwise.
+func (m *CgroupManager) GetParentPodMetadata(cgroupId uint64) *PodMetadata {
```

Comment:
Is there a reason you didn't want to read from the map directly (behind a sync.RWMutex)? Seems like it might be easier.

Reply:
I made a conscious choice to process all the events via a channel, and didn't want to deal with locks at this time. The channel approach is more easy to follow, and no risk of forgetting to lock, or run into deadlocks, etc.